edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/AuthAgentProtocolHead.cs

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.AspNetCore;
+    using Microsoft.AspNetCore.Hosting;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Hub.Mqtt;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.DependencyInjection.Extensions;
+    using Microsoft.Extensions.Logging;
+
+    public class AuthAgentProtocolHead : IProtocolHead
+    {
+        readonly IAuthenticator authenticator;
+        readonly IUsernameParser usernameParser;
+        readonly IClientCredentialsFactory clientCredentialsFactory;
+        readonly AuthAgentProtocolHeadConfig config;
+        readonly object guard = new object();
+
+        Option<IWebHost> host;
+
+        public string Name => "AUTH";
+
+        public AuthAgentProtocolHead(
+                    IAuthenticator authenticator,
+                    IUsernameParser usernameParser,
+                    IClientCredentialsFactory clientCredentialsFactory,
+                    AuthAgentProtocolHeadConfig config)
+        {
+            this.authenticator = Preconditions.CheckNotNull(authenticator, nameof(authenticator));
+            this.usernameParser = Preconditions.CheckNotNull(usernameParser, nameof(usernameParser));
+            this.clientCredentialsFactory = Preconditions.CheckNotNull(clientCredentialsFactory, nameof(clientCredentialsFactory));
+            this.config = Preconditions.CheckNotNull(config);
+        }
+
+        public async Task StartAsync()
+        {
+            Events.Starting();
+
+            lock (this.guard)
+            {
+                if (this.host.HasValue)
+                {
+                    Events.StartedWhenAlreadyRunning();
+                    throw new InvalidOperationException("Cannot start AuthAgentProtocolHead twice");
+                }
+                else
+                {
+                    this.host = Option.Some(
+                                    CreateWebHostBuilder(
```

Comment:
Interesting that this works. So we essentially have 2 asp.net web servers in the same process. Any concerns about it conflicting or something like that?

Reply:
my only concern was that they find each others controllers, but the new controller doesn't have routing attribute and the 'old' would use that for config, and the new controller is set manually in the new server, so that doesn't search for anything else.
I figured it would be both harder to combine the two (the existing has specialized middlewares for example)